### PR TITLE
[Android] Changes to incorporate synchronized transaction commits

### PR DESCRIPTION
**Pull request by adrianknight89**
### Description of Change ###

Re-doing #451. `CommitNowAllowingStateLoss()` can be used for single transactions that need immediate execution where we have not added transactions to the back stack.

Ref: https://medium.com/@bherbst/the-many-flavors-of-commit-186608a015b1

See comments for further discussion.

### Bugs Fixed ###

- N/A

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Platform.Android/AppCompat/NavigationPageRenderer.cs [hunk 1]
```diff
@@ -583,7 +597,6 @@ void RemovePage(Page page)
 
 			// Go ahead and take care of the fragment bookkeeping for the page being removed
 			FragmentTransaction transaction = FragmentManager.BeginTransaction();
-			transaction.DisallowAddToBackStack();
```
Comment (adrianknight89) [inline Xamarin.Forms.Platform.Android/AppCompat/NavigationPageRenderer.cs]: This method call was removed in #919 for master-detail and navigation pages. Looks like this was missed. Let me know if this needs to go in another PR.

File: Xamarin.Forms.Platform.Android/AppCompat/MasterDetailContainer.cs [hunk 2]
```diff
@@ -61,24 +60,10 @@ protected override void AddChildView(VisualElement childView)
...
-					FragmentTransaction transaction = FragmentManager.BeginTransaction();
-					transaction.Remove(_currentFragment);
-					transaction.SetTransition((int)FragmentTransit.None);
-
-					// This is a removal of a fragment that's not going on the back stack; there's no reason to care
-					// whether its state gets successfully saved, since we'll never restore it. Ergo, CommitAllowingStateLoss
-					transaction.CommitAllowingStateLoss();
-
-					_currentFragment = null;
-				}
```
Comment (adrianknight89) [inline Xamarin.Forms.Platform.Android/AppCompat/MasterDetailContainer.cs]: This whole block can be refactored into a method since similar logic is being used while disposing the container. Note that this block was NOT calling `ExecutePendingTransactions()` whereas the refactored method IS. I'm not sure if this can create any issues. @hartez removed forced execution of pending transactions in #8771b4fa for navigation pages. Let me know if this needs to change back to the original logic.

File: Xamarin.Forms.Platform.Android/AppCompat/MasterDetailContainer.cs [hunk 3]
```diff
@@ -87,65 +72,27 @@ protected override void AddChildView(VisualElement childView)
...
 
 			if (disposing)
 			{
-				if (_currentFragment != null && !FragmentManager.IsDestroyed)
-				{
-					FragmentTransaction transaction = FragmentManager.BeginTransaction();
-					transaction.Remove(_currentFragment);
-					transaction.SetTransition((int)FragmentTransit.None);
-					transaction.CommitAllowingStateLoss();
-					FragmentManager.ExecutePendingTransactions();
```
Comment (adrianknight89) [inline Xamarin.Forms.Platform.Android/AppCompat/MasterDetailContainer.cs]: `ExecutePendingTransactions()` here was not being run inside a looper, but the refactored method is. Is this OK?

File: Xamarin.Forms.Platform.Android/AppCompat/MasterDetailContainer.cs [hunk 4]
```diff
@@ -160,5 +107,59 @@ public void SetFragmentManager(FragmentManager fragmentManager)
...
+			FragmentTransaction fragmentTransaction = FragmentManager.BeginTransaction();
+
+			if (_currentFragment != null)
+				fragmentTransaction.Remove(_currentFragment);
+
+			fragmentTransaction.Add(Id, fragment);
+			fragmentTransaction.SetTransition((int)FragmentTransit.None);
+
+			// The view we're hosting in the fragment was never created (possibly we're already 		
+			// navigating to another page?) so there's nothing to commit synchronously.
```
Comment (adrianknight89) [inline Xamarin.Forms.Platform.Android/AppCompat/MasterDetailContainer.cs]: I added the word `synchronously` to the comment. I believe that's what was meant since we're not executing pending transactions?

**Comment by rmarinho** [comment]
Hi @adrianknight89 failed a lot of Android UITEsts, special api 23 and api 19, does it work fine locally?

**Comment by adrianknight89** [comment]
It worked locally for me, and I ran the first 250 tests successfully. It takes a long time for the tests to finish so I had to stop. However, I did my testing on API 25. Do you trigger UI tests on multiple API levels for each pull request?

**Comment by rmarinho** [comment]
@adrianknight89 yes we now run 2 sets of UITests for 3 API's on Android , normal renderers and fast renderers, on API19, API23, API25, on iOS we uitest against iOS9, iOS10 and iOS11.

We have some alpha uitests local runs for UWP and MacOS but they aren't integrated yet.

**Comment by rmarinho** [comment]
@adrianknight89 also XTC sometimes just fails. and we re run manually to see if it's a PR related issue or just network or XTC.. they are re-running now to check new results

**Comment by adrianknight89** [comment]
I tested mine on S6. I'll test it on 19 and 23 later tonight.

**Comment by adrianknight89** [comment]
@rmarinho Not sure what happened to my workspace, but I'm getting a lot of compilation issues that fail to recognize .NET Standard 2.0. For example, when I open project properties, it's not available in the "Target Framework" dropdown. I verified that I have .NET Core 2.0 SDK installed and new VS projects do recognize Net Standard 2.0.

**Comment by rmarinho** [comment]
build

File: Xamarin.Forms.Platform.Android/AppCompat/NavigationPageRenderer.cs (hunk 1 above)
Comment (hartez) [reply Xamarin.Forms.Platform.Android/AppCompat/NavigationPageRenderer.cs]: This is fine - we should not be using DisallowAddToBackStack anywhere.

File: Xamarin.Forms.Platform.Android/AppCompat/MasterDetailContainer.cs [hunk 5]
```diff
@@ -160,5 +107,59 @@ public void SetFragmentManager(FragmentManager fragmentManager)
...
+		void ClearCurrentFragment()
+		{
+			if (_currentFragment == null)
+				return;
+
+			FragmentTransaction fragmentTransaction = FragmentManager.BeginTransaction();
+			fragmentTransaction.Remove(_currentFragment);
+			fragmentTransaction.SetTransition((int)FragmentTransit.None);
+
+			// This is a removal of a fragment that's not going on the back stack; there's no reason to care
```
Comment (hartez) [inline Xamarin.Forms.Platform.Android/AppCompat/MasterDetailContainer.cs]: This comment is now confusing here - some version of it should probably be in CommitTransaction.

**Comment by rmarinho** [comment]
can you rebase @adrianknight89 ? need to fix the cert issue

**Comment by rmarinho** [comment]
@adrianknight89 any chance to rebase ?

**Comment by adrianknight89** [comment]
@rmarinho Sorry, I was away from my workspace for a while. Should be done.

**Comment by rmarinho** [comment]
thanks @adrianknight89

**Comment by rmarinho** [comment]
failing a lot of tests

**Comment by adrianknight89** [comment]
Not sure why that’s happening, and I don’t have time to work on this now. I’ll close this and open an enhancement ticket so the team works on it in the future.
